### PR TITLE
[Merged by Bors] - refactor: move internals from `entity_ref` to `World`, add `SAFETY` comments

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

There are some utility functions for actually working with `Storages` inside `entity_ref.rs` that are used both for `EntityRef/EntityMut` and `World`, with a `// TODO: move to Storages`.
This PR moves them to private methods on `World`, because that's the safest API boundary. On `Storages` you would need to ensure that you pass `Components` from the same world.

## Solution

- move get_component[_with_type], get_ticks[_with_type], get_component_and_ticks[_with_type] to `World` (still pub(crate))
- replace `pub use entity_ref::*;` with `pub use entity_ref::{EntityRef, EntityMut}` and qualified `entity_ref::get_mut[_by_id]` in `world.rs`
- add safety comments to a bunch of methods

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
I agree with the code reorg. Lemme take a look at the safety comments...

File: crates/bevy_ecs/src/storage/mod.rs [hunk 1]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
     pub tables: Tables,
     pub resources: Resources,
 }
+
+impl Storages {
+    /// Get a raw pointer to a particular [`Component`](crate::component::Component) and its [`ComponentTicks`] identified by their [`TypeId`]
+    ///
+    /// # Safety
+    /// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside
+    /// - `component_id` must be valid
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:33]: We're passing in a `type_id` here, so I don't think this applies.

File: crates/bevy_ecs/src/storage/mod.rs [hunk 2]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
     pub sparse_sets: SparseSets,
     pub tables: Tables,
     pub resources: Resources,
 }
+
+impl Storages {
+    /// Get a raw pointer to a particular [`Component`](crate::component::Component) and its [`ComponentTicks`] identified by their [`TypeId`]
+    ///
+    /// # Safety
+    /// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:32]: A specific archetype isn't given, so this comment is unclear.

File: crates/bevy_ecs/src/storage/mod.rs [hunk 3]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
+        location: EntityLocation,
+    ) -> Option<(Ptr<'_>, &UnsafeCell<ComponentTicks>)> {
+        let component_id = components.get_id(type_id)?;
+        self.get_component_and_ticks(archetypes, components, component_id, entity, location)
+    }
+
+    /// Get a raw pointer to a particular [`Component`](crate::component::Component) and its [`ComponentTicks`]
+    ///
+    /// # Safety
+    /// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:52]: Ditto for "given archetype". This seems to be a problem on pretty much every method here :)

File: crates/bevy_ecs/src/storage/mod.rs [hunk 4]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
+    ) -> Option<(Ptr<'_>, &UnsafeCell<ComponentTicks>)> {
+        let archetype = &archetypes[location.archetype_id];
+        // SAFETY: component_id is valid as per safety requirements
+        let component_info = unsafe { components.get_info_unchecked(component_id) };
+        match component_info.storage_type() {
+            StorageType::Table => {
+                let table = &self.tables[archetype.table_id()];
+                let components = table.get_column(component_id)?;
+                let table_row = archetype.entity_table_row(location.index);
+                // SAFETY: archetypes only store valid table_rows and the stored component type is T
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:73]: It's not clear what `T` refers to in the context of this function.

File: crates/bevy_ecs/src/storage/mod.rs [hunk 5]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
+    ) -> Option<(Ptr<'_>, &UnsafeCell<ComponentTicks>)> {
+        let component_id = components.get_id(type_id)?;
+        self.get_component_and_ticks(archetypes, components, component_id, entity, location)
+    }
+
+    /// Get a raw pointer to a particular [`Component`](crate::component::Component) and its [`ComponentTicks`]
+    ///
+    /// # Safety
+    /// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside
+    /// - `component_id` must be valid
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:53]: I'd like a bit more advice here on what it means for a component id to be valid. Probably best suited to living on the `ComponentId` docs though.

File: crates/bevy_ecs/src/storage/mod.rs [hunk 6]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
+    ) -> Option<&UnsafeCell<ComponentTicks>> {
+        let archetype = &archetypes[location.archetype_id];
+        // SAFETY: component_id is valid as per safety requirements
+        let component_info = unsafe { components.get_info_unchecked(component_id) };
+        match component_info.storage_type() {
+            StorageType::Table => {
+                let table = &self.tables[archetype.table_id()];
+                let components = table.get_column(component_id)?;
+                let table_row = archetype.entity_table_row(location.index);
+                // SAFETY: archetypes only store valid table_rows and the stored component type is T
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:188]: T is not defined here.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Okay, so the safety docs have some problems 😂 They're pretty much all repeated, and I didn't call them out in each location to reduce spam.

Not your fault by any means (these docs are much better than we had), but eh, I'd like to clean them up while we're here.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 7]
```diff
@@ -191,9 +231,21 @@ impl<'w> EntityMut<'w> {
...
+        // SAFETY:
+        // - entity location is valid
+        // - archetypes and components come from the same world
+        // - world access is immutable, lifetime tied to `&self`
         unsafe {
-            get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
+            self.world
+                .storages
+                .get_component_with_type(
+                    &self.world.archetypes,
```
Comment (james7132) [inline crates/bevy_ecs/src/world/entity_ref.rs:242]: Almost all of these are using borrows from `World`. Wouldn't it make more sense to move these to `World` than `Storages`?

File: crates/bevy_ecs/src/storage/mod.rs [hunk 8]
```diff
@@ -9,10 +9,235 @@ pub use resource::*;
...
+impl Storages {
+    /// Get a raw pointer to a particular [`Component`](crate::component::Component) and its [`ComponentTicks`] identified by their [`TypeId`]
+    ///
+    /// # Safety
+    /// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside
+    /// - `component_id` must be valid
+    /// - `Archetypes` and `Components` must come from the world this of this `Storages`
+    /// - the caller must ensure that no aliasing rules are violated
+    #[inline]
+    pub unsafe fn get_component_and_ticks_with_type(
```
Comment (james7132) [inline crates/bevy_ecs/src/storage/mod.rs:37]: These were `pub(crate)` before, why are they now just `pub`?

File: crates/bevy_ecs/src/world/entity_ref.rs (hunk 7 above)
Comment (james7132) [reply crates/bevy_ecs/src/world/entity_ref.rs:243]: If it cannot be safely used without them, it's best to put it on the interface where it's most cleanly used and has fewer safety invariants.

File: crates/bevy_ecs/src/storage/mod.rs (hunk 8 above)
Comment (james7132) [reply crates/bevy_ecs/src/storage/mod.rs:36]: I'd err on the side of caution here. Unless we have a solid external use case, having more `pub`, particularly for items that mutably change the state of the `World`, is just asking for public facing unsoundness.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 9]
```diff
@@ -666,13 +664,13 @@ impl<'w> EntityMut<'w> {
     #[inline]
     pub fn get_by_id(&self, component_id: ComponentId) -> Option<Ptr<'_>> {
         let info = self.world.components().get_info(component_id)?;
+        self.world.components().get_info(component_id)?;
```
Comment (james7132) [inline crates/bevy_ecs/src/world/entity_ref.rs:667]:
```suggestion
```

This is redundant with the above.

File: crates/bevy_ecs/src/world/mod.rs [hunk 10]
```diff
@@ -1547,6 +1549,170 @@ impl World {
...
+                // SAFETY: archetypes only store valid table_rows and caller ensure aliasing rules
+                Some(components.get_ticks_unchecked(table_row))
+            }
+            StorageType::SparseSet => fetch_sparse_set(self, component_id)?.get_ticks(entity),
+        }
+    }
+}
+
+#[inline]
+unsafe fn fetch_table(
```
Comment (james7132) [inline crates/bevy_ecs/src/world/mod.rs:1699]: Any reason why these couldn't be in the block above?

**Review by james7132 — CHANGES_REQUESTED** [review]

**Review by james7132 — APPROVED** [review]

File: crates/bevy_ecs/src/world/mod.rs [hunk 11]
```diff
@@ -1547,6 +1549,170 @@ impl World {
...
+                let (components, table_row) = self.fetch_table(location, component_id)?;
+                // SAFETY: archetypes only store valid table_rows and caller ensure aliasing rules
+                Some(components.get_ticks_unchecked(table_row))
+            }
+            StorageType::SparseSet => self.fetch_sparse_set(component_id)?.get_ticks(entity),
+        }
+    }
+
+    #[inline]
+    unsafe fn fetch_table(
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1698]: This needs safety comments.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 12]
```diff
@@ -1071,16 +863,65 @@ pub(crate) unsafe fn get_mut_by_id(
...
+/// returned pointer is semantically owned by the caller and will not be dropped in its original location.
+/// Caller is responsible to drop component data behind returned pointer.
+///
+/// # Safety
+/// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside the `archetype`
+/// - `component_id` must be valid
+/// - `components`, `archetypes` and `storages` must come from the same world
+/// - The relevant table row **must be removed** by the caller once all components are taken
+#[inline]
+pub(crate) unsafe fn take_component<'a>(
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:895]: The PR description suggests that this is a method on Storages (which I like better), but here it's a freestanding function.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Two concerns, but once those are fixed up I'm happy to merge this. It's better organized, and the safety comments are much better.

**Review by alice-i-cecile — APPROVED** [review]
LGTM now once the PR description matches where the methods were moved, one way or another.

**Pull request closed without merging**